src/backend/distributed/operations/auto_shard_split.c

```diff
@@ -0,0 +1,370 @@
+ * If a tenant is found splitpoints for isolation is returned otherwise average hash value is returned.
+ */
+List *
+FindShardSplitPoints(ShardInfo shardinfo)
+{
+	StringInfo CommonValueQuery = makeStringInfo();
+
+	/*
+	 * The inner query for extracting the tenant value having frequency > 0.3 is executed on
+	 * every shard of the table and outer query gives the shardid and tenant values as the
+	 * output.
+	 */
+	appendStringInfo(CommonValueQuery,
+					 "SELECT shardid , unnest(result::%s[]) from run_command_on_shards(%s,$$SELECT array_agg(val)"
+					 " FROM pg_stats s , unnest(most_common_vals::text::%s[],most_common_freqs) as res(val,freq)"
+					 " WHERE tablename = %s AND attname = %s AND freq > 0.2 $$)"
+					 " WHERE result <> '' AND shardid = %ld;",
+					 shardinfo->datatype, quote_literal_cstr(shardinfo->tablename),
+					 shardinfo->datatype,
+					 quote_literal_cstr(shardinfo->shardname),
+					 quote_literal_cstr(shardinfo->distributionColumn),
+					 shardinfo->shardid);
+
+	ereport(LOG, errmsg("%s", CommonValueQuery->data));
+	List *SplitPoints = NULL;
+	MemoryContext originalContext = CurrentMemoryContext;
+	SPI_connect();
+	SPI_exec(CommonValueQuery->data, 0);
+	MemoryContext spiContext = CurrentMemoryContext;
+	int64 rowCount = SPI_processed;
+	int64 average;
+	int32 hashedValue;
+	ereport(LOG, errmsg("%ld", rowCount));
+
+	if (rowCount > 0)
+	{
+		SPITupleTable *tupletable = SPI_tuptable;
+		CitusTableCacheEntry *cacheEntry = GetCitusTableCacheEntry(shardinfo->tableId);
+		for (int rowIndex = 0; rowIndex < rowCount; rowIndex++)
+		{
+			HeapTuple tuple = tupletable->vals[rowIndex];
+			char *commonValue = SPI_getvalue(tuple, tupletable->tupdesc, 2);
+			ereport(LOG, errmsg("%s", commonValue));
+			Datum tenantIdDatum = StringToDatum(commonValue,
+												shardinfo->distributionColumnId);
+			Datum hashedValueDatum = FunctionCall1Coll(cacheEntry->hashFunction,
+													   cacheEntry->partitionColumn->
+													   varcollid,
+													   tenantIdDatum);
+			hashedValue = DatumGetInt32(hashedValueDatum);
+			ereport(LOG, errmsg("%d", hashedValue));
+			/*Switching the memory context to store the unique SplitPoints in a list*/
+
+			MemoryContextSwitchTo(originalContext);
+			if (hashedValue == shardinfo->shardminvalue)
+			{
+				SplitPoints = list_append_unique_int(SplitPoints,hashedValue);
+			}
+			else if (hashedValue == shardinfo->shardmaxvalue)
+			{
+				SplitPoints = list_append_unique_int(SplitPoints, hashedValue - 1);
+			}
+			else
+			{
+				SplitPoints = list_append_unique_int(SplitPoints, hashedValue - 1);
+				SplitPoints = list_append_unique_int(SplitPoints, hashedValue);
+			}
+			MemoryContextSwitchTo(spiContext);
+		}
+		SPI_freetuptable(tupletable);
+	}
+	else
+	{
+		average = ExecuteAvgHashQuery(shardinfo);
+		ereport(LOG, errmsg("%ld", average));
+	}
+
+	SPI_finish();
+
+	if (rowCount > 0)
+	{
+		list_sort(SplitPoints, list_int_cmp);
```

Comment:
I was wondering why we needed this, but then realized we're basically doing multiple tenant isolations at once. That's kind of neat.